### PR TITLE
[Merged by Bors] - feat: port Data.MvPolynomial.Funext

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/MvPolynomial/Funext.lean [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  over an infinite integral domain `R` are equal if `eval x φ = eval x ψ` for all `x : σ → R`.
+
+-/
+
+
+namespace MvPolynomial
+
+variable {R : Type _} [CommRing R] [IsDomain R] [Infinite R]
+
+set_option maxHeartbeats 1000000 in
```
Comment (Parcly-Taxel) [inline Mathlib/Data/MvPolynomial/Funext.lean:33]: If this line is absent, there are timeouts. Should I leave it this way?

File: Mathlib/Data/MvPolynomial/Funext.lean [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    convert h finZeroElim
+    /- Porting note: this branch of the proof closes here in Lean 4!
+    suffices
+      (eval₂_hom (RingHom.id _) (IsEmpty.elim' Fin.isEmpty)) p =
+        (eval finZeroElim : MvPolynomial (Fin 0) R →+* R) p by
+      rw [← this]
+      simp only [coe_eval₂Hom, is_empty_ring_equiv_apply, RingEquiv.trans_apply,
+        aeval_eq_eval₂Hom]
+      congr
+    exact eval₂Hom_congr rfl (Subsingleton.elim _ _) rfl -/
```
Comment (semorrison) [inline Mathlib/Data/MvPolynomial/Funext.lean:49]: I'd say we can just drop this. Very important to leave porting notes for any regressions, but when Lean 4 is just better, it's okay to simply clean up.

File: Mathlib/Data/MvPolynomial/Funext.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/Data/MvPolynomial/Funext.lean:33]: At minimum, needs to have a `-- Porting note`

Comment (semorrison) [reply Mathlib/Data/MvPolynomial/Funext.lean:33]: I'm trying to look at this now, but waiting on oleans / `cache get!` problems.

Two possibilities that I'd be okay with:
1) Leave it in, with a porting note
2) Split out two separate lemmas `funext_fin_zero` and `funext_fin_succ`.

Comment (semorrison) [reply Mathlib/Data/MvPolynomial/Funext.lean:33]: Okay, the answer here is that this proof should never have been in mathlib3 in the first place. There is a lemma that should have been factored out:

```
theorem eval_eval_finSuccEquiv (f : MvPolynomial (Fin (n + 1)) R) :
    (eval x) (Polynomial.eval q (finSuccEquiv R n f)) = eval (Fin.cons (eval x q) x) f := sorry

private theorem funext_fin {n : ℕ} {p : MvPolynomial (Fin n) R}
    (h : ∀ x : Fin n → R, eval x p = 0) : p = 0 := by
  induction' n with n ih
  · apply (MvPolynomial.isEmptyRingEquiv R (Fin 0)).injective
    rw [RingEquiv.map_zero]
    convert h finZeroElim
  · apply (finSuccEquiv R n).injective
    simp only [AlgEquiv.map_zero]
    refine Polynomial.funext fun q => ?_
    rw [Polynomial.eval_zero]
    apply ih fun x => ?_
    calc _ = _ := eval_eval_finSuccEquiv p
         _ = 0 := h _
```

`eval_eval_finSuccEquiv` needs to be proved, and moved to `Mathlib.Data.MvPolynomial.Equiv`, after `eval_eq_eval_mv_eval'`.

In @jcommelin's defence, when he wrote this proof in https://github.com/leanprover-community/mathlib/pull/4196 the API in `Mathlib.Data.MvPolynomial.Equiv` wasn't as developed.

@Parcly-Taxel, do you want to tackle this? (Up to you of course.)

Comment (semorrison) [reply Mathlib/Data/MvPolynomial/Funext.lean]: @Parcly-Taxel, I went down the rabbit hole and proved the theorem. There are some `TODO`s about things that need to be moved, still.

Comment (Parcly-Taxel) [reply Mathlib/Data/MvPolynomial/Funext.lean]: @semorrison Should we make a mathlib3 PR that moves the theorems to their proper files and rewrites `funext_fin` accordingly?

Comment (eric-wieser) [reply Mathlib/Data/MvPolynomial/Funext.lean]: I think that would be a great idea, if you're up for it

Comment (semorrison) [reply Mathlib/Data/MvPolynomial/Funext.lean]: I've done this now.

**Comment by semorrison** [comment]
As I've now written most of this, can someone else please review and merge?

File: Mathlib/Data/MvPolynomial/Basic.lean [hunk 3]
```diff
@@ -1148,6 +1148,16 @@ theorem eval_assoc {τ} (f : σ → MvPolynomial τ R) (g : τ → R) (p : MvPol
   congr with a; simp
 #align mv_polynomial.eval_assoc MvPolynomial.eval_assoc
 
+theorem eval_eval₂ [CommSemiring R] [CommSemiring S]
```
Comment (jcommelin) [inline Mathlib/Data/MvPolynomial/Basic.lean:1151]:
```suggestion
-- Porting note: new theorem
theorem eval_eval₂ [CommSemiring R] [CommSemiring S]
```

File: Mathlib/Data/MvPolynomial/Basic.lean [hunk 4]
```diff
@@ -1204,6 +1214,11 @@ theorem eval₂_eq_eval_map (g : σ → S₁) (p : MvPolynomial σ R) : p.eval
     simp only [comp_apply, eval₂_X]
 #align mv_polynomial.eval₂_eq_eval_map MvPolynomial.eval₂_eq_eval_map
 
+-- This probably belongs earlier, but it breaks the fragile proof of `eval₂_eq_eval_map`
```
Comment (jcommelin) [inline Mathlib/Data/MvPolynomial/Basic.lean:1217]:
```suggestion
-- Porting note: new theorem
-- This probably belongs earlier, but it breaks the fragile proof of `eval₂_eq_eval_map`
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**